### PR TITLE
Remove `error()`, expose `printError()`

### LINKED ISSUE #178: TypeScript doesn't know that error() is terminal
TypeScript should be able to detect that using Waterfall CLI's `error()` function is terminal.

Currently, this function is returning a 'void' type, and I guess it should be 'never', so that the compiler will understand that program execution does not continue after being called.

To get around this issue currently, I typically have to put some unreachable but terminal (eg: `process.exit(1);`) after the call to `error()` so that TypeScript will understand the finality of this code-path.

### LINKED ISSUE #334: Remove `error()`, add `printError()`
In order to avoid some issues to do with `error()` being called from sync vs async contexts, we've decided to unbundle what `error()` used to do, and make it just be about printing an error message in a standardized way, and to make the calling side responsible for what else happens in response to an error condition having been detected.

ie: 

- flushing of `stdout` before printing the error to `stderr`  
- calling `process.exit(255)` (which, used within async context, would *not* immediately cause program termination)

### PR DESCRIPTION
**Pull request by pbelbin**
Closes #334  
Closes #178 

Remove `error()` and expose `printError()` instead.

This means the user of `printError()` must also take some precautions regarding potential `stdout` content that has yet to be flushed, and, what to do *after* printing the error message.

### CODE REVIEW

**Comment by pbelbin** [comment]
I'm wondering....

Looking at how the clients of `waterfall-cli` will be impacted by this change, would it be helpful / useful if the `printError()` happened to return the string that was sent to `stderr`?

In cases where the text is wanted for use within a subsequently thrown `Error`, it might be convenient, so it could look a bit like this, perhaps:

```
.
.
throw new Error(printError('error message to be output`));
.
.
```

Perhaps just the text would be output, and not the 'ERROR' banner?

**Comment by WesCossick** [comment]
> Perhaps just the text would be returned, and not the 'ERROR' banner?

Returning the error sounds like a good idea! But yeah, I'd definitely only return the plain text that was given to the function, not the stylized version.

**Comment by pbelbin** [comment]
I kinda wonder if being able to do this would be helpful, or not:

```
import { WaterfallError } from `waterfall-cli`

.
.
throw new WaterfallError('message text here');
.
.
```

and within, it would *also* do the `printError()` call?

I suppose it's comparison would be:
```
throw new Error(waterfall.printError('message text here'));
```

Not sure how much of a difference it would really be, but wanted to mention it.

It looks a little bit less complex to look at, but it hides a bit what's happening.

If the error type was a member of waterfall, I suppose it would look like:

```
import waterfall from 'waterfall-cli';

.
.
throw new waterfall.Error('message text here');
.
.
```

**Comment by WesCossick** [comment]
Yeah, I'm not sure if that'd be useful. Something to consider for the future.

File: src/print-pretty-error.ts [hunk 1]
```diff
@@ -1,8 +1,10 @@
 // Dependencies
 import chalk from './chalk';
 
-// Print a pretty error message
+// Print a pretty error message and return just the message
 export default function printPrettyError(message: string) {
 	console.error(`${chalk.inverse.red.bold(' ERROR ')}\n`);
 	console.error(chalk.red(`> ${message.split('\n').join('\n> ')}\n`));
+
+	return message;
```
Comment (WesCossick) [inline src/print-pretty-error.ts:9]: See warning from Codecov below. This line needs testing.

**Review by WesCossick — CHANGES_REQUESTED** [review]
See comment

File: src/print-pretty-error.ts (hunk 1 above)
Comment (pbelbin) [reply src/print-pretty-error.ts:9]: I added a test that checks the response value.

**Review by pbelbin — COMMENTED** [review]
A test was added that specifically checks the return value from print-pretty-error via the exposed `printError()` response being checked.

**Review by pbelbin — COMMENTED** [review]
Actually, the additional test was added to `print-pretty-error.test.ts`

**Comment by pbelbin** [comment]
hey @WesCossick , please review.  This needs to be merged in and released before the dependent project work can be completed.

File: test/print-pretty-error.test.ts [hunk 2]
```diff
@@ -42,4 +42,8 @@ describe('#printPrettyError()', () => {
 
 		expect(consoleResult).toContain('Error message content');
 	});
+
+	test('Returns the message text', () => {
+		expect(printPrettyError('Error message content')).toContain('Error message content');
```
Comment (WesCossick) [inline test/print-pretty-error.test.ts:47]: This is a stricter check:

```suggestion
		expect(printPrettyError('Error message content')).toEqual('Error message content');
```

**Review by WesCossick — CHANGES_REQUESTED** [review]

**Review by WesCossick — APPROVED** [review]

**Pull request merged**